ls/src/file.rs

```diff
@@ -0,0 +1,236 @@
+use coreutils_core::os::{group::Group, passwd::Passwd};
+
+use std::{
+    fs, io,
+    os::unix::fs::{FileTypeExt, MetadataExt, PermissionsExt},
+    path,
+    string::String,
+};
+
+use ansi_term::Color;
+
+extern crate chrono;
+
+use chrono::{DateTime, Local};
+
+use crate::flags::Flags;
+
+/// Represents a file and it's properties
+pub(crate) struct File {
+    pub name: String,
+    pub path: path::PathBuf,
+    pub metadata: fs::Metadata,
+    flags: Flags,
+}
+
+impl File {
+    /// Creates a `File` instance from a `DirEntry`
+    pub fn from(path: path::PathBuf, flags: Flags) -> io::Result<Self> {
+        let metadata = path.symlink_metadata().expect("Failed to read metadata?");
+
+        if flags.dereference && metadata.file_type().is_symlink() {
+            let symlink = fs::read_link(path.clone())?;
+
+            let name: String = File::path_to_file_name(&symlink);
+
+            let metadata = path.metadata().unwrap();
+
+            return Ok(File { name, path: symlink, metadata, flags });
+        }
+
+        let name = File::path_to_file_name(&path);
+
+        Ok(File { name, path, metadata, flags })
+    }
+
+    /// Creates a `File` instance from a `DirEntry` and supplies a file name
+    pub fn from_name(name: String, path: path::PathBuf, flags: Flags) -> io::Result<Self> {
+        let metadata = path.metadata().expect("Failed to read metadata");
+
+        Ok(File { name, path, metadata, flags })
+    }
+
+    /// Retrieves the number of blocks allocated to a file as a string
+    pub fn get_blocks(&self) -> String { self.metadata.blocks().to_string() }
+
+    /// Retrieves a files permissions as a string
+    pub fn get_permissions(&self) -> String {
+        let mode = self.metadata.permissions().mode();
+
+        unix_mode::to_string(mode)
+    }
+
+    /// Retrieves the number of hard links pointing to a file as a string
+    pub fn get_hard_links(&self) -> String { self.metadata.nlink().to_string() }
+
+    pub fn get_inode(&self) -> String { self.metadata.ino().to_string() }
+
+    /// Retrieves the file's user name as a string. If the `-n` flag is set,
+    /// the the user's ID is returned
+    pub fn get_user(&self) -> String {
+        let user: String;
+
+        if self.flags.numeric_uid_gid {
+            let user_value = self.metadata.uid();
+            user = user_value.to_string();
+        } else {
+            let uid = Passwd::from_uid(self.metadata.uid()).unwrap();
```

Comment:
I noticed that the GNU implementation returns the `uid` if it's unable to get the owners name. Would it be acceptable to use `unwrap_or_else()` (see below), rather than returning a `Result` in this instance?

``` rust
Passwd::from_uid(self.metadata.uid()).unwrap_or_else(|| self.metadata.uid());
```
***
 PS: sorry for all the panicking. I am still getting used to Rust's error handling.